### PR TITLE
Reimplement script operands using OOP

### LINKED ISSUE #160: Lua error when setting a direct string value in a condition
Since 1.3 I cannot set direct string values anymore. The textbox will not show and I get a Lua error (message below).

My steps:

1. create a new campaign
2. switch to "Actions" tab
3. add any new action
4. CTRL+click to add a condition
5. click on the default condition, prefilled with "Unit name (target) is equal to Elsa"
6. on the right side (value) open the menu and choose Direct Value > String Value

It seems that the value passed to SetText() is a table instead of string.

`Message: ...talRP3_Extended_Tools\script/conditions/operands.lua:67: Usage: <unnamed>:SetText("text")
Time: Sat Jan 12 13:32:22 2019
Count: 3
Stack: ...talRP3_Extended_Tools\script/conditions/operands.lua:67: Usage: <unnamed>:SetText("text")
[C]: in function `SetText'
...talRP3_Extended_Tools\script/conditions/operands.lua:67: in function `load'
...lRP3_Extended_Tools\script/conditions/conditions.lua:135: in function <...lRP3_Extended_Tools\script/conditions/conditions.lua:123>
...lRP3_Extended_Tools\script/conditions/conditions.lua:178: in function <...lRP3_Extended_Tools\script/conditions/conditions.lua:157>
...lRP3_Extended_Tools\script/conditions/conditions.lua:653: in function <...lRP3_Extended_Tools\script/conditions/conditions.lua:651>
Interface\AddOns\totalRP3\core\impl\ui_tools.lua:254: in function <Interface\AddOns\totalRP3\core\impl\ui_tools.lua:245>
Interface\AddOns\totalRP3\core\impl\ui_tools.lua:169: in function `func'
...3\libs\MSA-DropDownMenu-1.0\MSA-DropDownMenu-1.0.lua:1183: in function `MSA_DropDownMenuButton_OnClick'
...3\libs\MSA-DropDownMenu-1.0\MSA-DropDownMenu-1.0.lua:156: in function <...3\libs\MSA-DropDownMenu-1.0\MSA-DropDownMenu-1.0.lua:155>

Locals: (*temporary) = <unnamed> {
 0 = <userdata>
 Center = <unnamed> {
 }
 help = TRP3_OperandEditorStringHelp {
 }
 Right = <unnamed> {
 }
 Left = <unnamed> {
 }
 title = TRP3_OperandEditorStringText {
 }
}
(*temporary) = <table> {
}`

### PR DESCRIPTION
**Pull request by Ellypse**
This re-implements the entire script operands system to use object-oriented programming instead of simple tables.

The `Operand` and `NumericOperand` classes should be used to create a new operand. The newly created `Operand` will register itself with the script operands system. It expects a unique ID and will throw an error if an `Operand` is being created with an existing ID. An environment can be provided when the `Operand` is created, otherwise, the Lua environment in which the `Operand` is executed will be empty.

An `Operand` SHOULD redefine the `Operand:CodeReplacement(args)` method to return the generated code that should be used for the operand, using the arguments passed to the method.

The code has been updated to use the new OOP notation, and temporary backward compatibility has been added to support calling the old `Operand.codeRemplacement(args)` in the previous code.

Some basic unit tests have been added to the project. The first tests can be run by executing the global function `__TRP3ExtendedTests__operands()` and will check that all registered operands code generation does not error when given unexpected arguments.

Fixes #160

### CODE REVIEW

File: totalRP3_Extended/script/Operands/Operands.lua [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+--- Retrieve a specific desired argument from the arguments passed to an operand in a safe way.
+--- The arguments table is checked to be valid and that the desired argument has been provided, otherwise the default
+--- value provided is returned.
+---@generic T
+---@param arguments table|nil The arguments given to the operand
+---@param argumentIndex number|string The index where the desired argument should be stored in the arguments table
+---@param defaultValue T A default value that should be returned if the desired argument could not be retrieved.
+---@return T Either the value passed via the arguments if found or the default value
+function TRP3_API.getSafeValueFromTable(arguments, argumentIndex, defaultValue)
```
Comment (Ellypse) [inline totalRP3_Extended/script/Operands/Operands.lua]: Note: This function is where we make sure we always retrieve a valid value out of the `args` passed to `Operand:CodeReplacement(args)`.

File: totalRP3_Extended/script/Operands/ExpertOperands.lua [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+	local var = getSafe(args, 2, 1);
+	return ([[eventVarCheck(args, %s)]]):format(var);
+end
+
+local checkEventNumericVariableValueOperand = NumericOperand("check_event_var_n", {
+	["eventVarCheckN"] = "TRP3_API.script.eventVarCheckN"
+});
+
+function checkEventNumericVariableValueOperand:CodeReplacement(args)
+	local var = getSafe(args, 2, 1);
```
Comment (Solanya) [inline totalRP3_Extended/script/Operands/ExpertOperands.lua]:
```suggestion
	local var = getSafe(args, 1, 1);
```

File: totalRP3_Extended/script/Operands/ExpertOperands.lua [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+	return ([[eventVarCheckN(args, %s)]]):format(var);
+end
+
+local randomOperand = Operand("random", {
```
Comment (Solanya) [inline totalRP3_Extended/script/Operands/ExpertOperands.lua]:
```suggestion
local randomOperand = NumericOperand("random", {
```

File: totalRP3_Extended/script/Operands/InventoryOperands.lua [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	end
+	return ([[getItemCount("%s", %s)]]):format(id, source);
+end
+
+local itemWeightOperand = NumericOperand("inv_item_weight", {
+	["getContainerWeight"] = "TRP3_API.inventory.getContainerWeight"
+});
+
+function itemWeightOperand:CodeReplacement(args)
+	local source = getSafe(args, 2, "nil");
```
Comment (Solanya) [inline totalRP3_Extended/script/Operands/InventoryOperands.lua]:
```suggestion
	local source = getSafe(args, 1, "nil");
```

File: totalRP3_Extended/script/Operands/UnitOperands.lua [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+----------------------------------------------------------------------------------
+
+---@type TRP3_API
+local TRP3_API = TRP3_API
+
+---@type TotalRP3_Extended_Operand
+local Operand = TRP3_API.script.Operand;
+---@type TotalRP3_Extended_NumericOperand
+local NumericOperand = TRP3_API.script.NumericOperand;
+local getSafeArgument = TRP3_API.getSafeValueFromTable;
```
Comment (Solanya) [inline totalRP3_Extended/script/Operands/UnitOperands.lua]:
```suggestion
local getSafe = TRP3_API.getSafeValueFromTable;
```

File: totalRP3_Extended/script/Operands/UnitOperands.lua [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+local function getUnitId(args)
+	return getSafeArgument(args, 1, "target");
```
Comment (Solanya) [inline totalRP3_Extended/script/Operands/UnitOperands.lua]:
```suggestion
	return getSafe(args, 1, "target");
```

File: totalRP3_Extended/script/Operands/UnitOperands.lua [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+local unitIsInInspectingDistanceOperand = Operand("unit_distance_inspect", {
+	["CheckInteractDistance"] = "CheckInteractDistance"
+});
+
+function unitIsInInspectingDistanceOperand:CodeReplacement(args)
+	return ([[CheckInteractDistance("%s", 1)]]):format(getUnitId(args));
+end
+
+local unitDistanceToPointOperand = Operand("unit_distance_point", {
```
Comment (Solanya) [inline totalRP3_Extended/script/Operands/UnitOperands.lua]:
```suggestion
local unitDistanceToPointOperand = NumericOperand("unit_distance_point", {
```

File: totalRP3_Extended/script/Operands/UnitOperands.lua [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+	["unitDistancePoint"] = "TRP3_API.extended.unitDistancePoint"
+});
+
+function unitDistanceToPointOperand:CodeReplacement(args)
+	local x = getSafeArgument(args, 3, 0)
+	local y = getSafeArgument(args, 2, 0)
+	return ([[unitDistancePoint("%s", %s, %s)]]):format(getUnitId(args), x, y);
+end
+
+local unitDistanceFromPlayerOperand = Operand("unit_distance_me", {
```
Comment (Solanya) [inline totalRP3_Extended/script/Operands/UnitOperands.lua]:
```suggestion
local unitDistanceFromPlayerOperand = NumericOperand("unit_distance_me", {
```

File: totalRP3_Extended/script/Operands/UnitOperands.lua [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+function unitIsInInspectingDistanceOperand:CodeReplacement(args)
+	return ([[CheckInteractDistance("%s", 1)]]):format(getUnitId(args));
+end
+
+local unitDistanceToPointOperand = Operand("unit_distance_point", {
+	["unitDistancePoint"] = "TRP3_API.extended.unitDistancePoint"
+});
+
+function unitDistanceToPointOperand:CodeReplacement(args)
+	local x = getSafeArgument(args, 3, 0)
```
Comment (Solanya) [inline totalRP3_Extended/script/Operands/UnitOperands.lua]:
```suggestion
	local x = getSafe(args, 3, 0)
```

File: totalRP3_Extended/script/Operands/UnitOperands.lua [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+	local y = getSafeArgument(args, 2, 0)
```
Comment (Solanya) [inline totalRP3_Extended/script/Operands/UnitOperands.lua]:
```suggestion
	local y = getSafe(args, 2, 0)
```

**Pull request merged**